tsdb/engine/tsm1/data_file.go

```diff
@@ -0,0 +1,666 @@
+	MagicNumber uint32 = 0x16D116D1
+
+	Version byte = 1
+
+	indexEntrySize = 28
+)
+
+// TSMWriter writes TSM formatted key and values.
+type TSMWriter interface {
+	// Write writes a new block for key containing and values.  Writes append
+	// blocks in the order that the Write function is called.  The caller is
+	// responsible for ensuring keys and blocks or sorted appropriately.
+	// Values are encoded as a full block.  The caller is responsible for
+	// ensuring a fixed number of values are encoded in each block as wells as
+	// ensuring the Values are sorted. The first and last timestamp values are
+	// used as the minimum and maximum values for the index entry.
+	Write(key string, values Values) error
+
+	// Close finishes the TSM write streams and writes the index.
+	Close() error
+}
+
+// TSMIndex represent the index section of a TSM file.  The index records all
+// blocks, their locations, sizes, min and max times.
+type TSMIndex interface {
+
+	// Add records a new block entry for a key in the index.
+	Add(key string, minTime, maxTime time.Time, offset int64, size uint32)
+
+	// Entries returns all index entries for a key.
+	Entries(key string) []*IndexEntry
+
+	// Entry returns the index entry for the specified key and timestamp.  If no entry
+	// matches the key an timestamp, nil is returned.
+	Entry(key string, timestamp time.Time) *IndexEntry
+
+	// MarshalBinary returns a byte slice encoded version of the index.
+	MarshalBinary() ([]byte, error)
+
+	// UnmarshalBinary populates an index from an encoded byte slice
+	// representation of an index.
+	UnmarshalBinary(b []byte) error
+}
+
+// IndexEntry is the index information for a given block in a TSM file.
+type IndexEntry struct {
+
+	// The min and max time of all points stored in the block.
+	MinTime, MaxTime time.Time
+
+	// The absolute position in the file where this block is located.
+	Offset int64
+
+	// The size in bytes of the block in the file.
+	Size uint32
+}
+
+func (e *IndexEntry) UnmarshalBinary(b []byte) error {
+	if len(b) != indexEntrySize {
+		return fmt.Errorf("unmarshalBinary: short buf: %v != %v", indexEntrySize, len(b))
+	}
+	e.MinTime = time.Unix(0, int64(btou64(b[:8])))
+	e.MaxTime = time.Unix(0, int64(btou64(b[8:16])))
+	e.Offset = int64(btou64(b[16:24]))
+	e.Size = btou32(b[24:28])
+	return nil
+}
+
+// Returns true if this IndexEntry may contain values for the given time.  The min and max
+// times are inclusive.
+func (e *IndexEntry) Contains(t time.Time) bool {
+	return e.MinTime.Equal(t) || e.MinTime.Before(t) &&
+		e.MaxTime.Equal(t) || e.MaxTime.After(t)
+}
+
+func NewDirectIndex() TSMIndex {
+	return &directIndex{
+		blocks: map[string]indexEntries{},
+	}
+}
+
+// directIndex is a simple in-memory index implementation for a TSM file.  The full index
+// must fit in memory.
+type directIndex struct {
+	blocks map[string]indexEntries
+}
+
+func (d *directIndex) Add(key string, minTime, maxTime time.Time, offset int64, size uint32) {
```

Comment:
It doesn't actually matter because keys are put in the `blocks` map so finding one doesn't require a search.  Finding a block for a timestamp also doesn't matter because we do a linear scan right now.

`Add` is only be called for building an index to write a new TSM file and sort the keys and timestamp blocks is done in `Close`.